src/analysis.ts

```diff
@@ -12,16 +12,309 @@
+    if (document instanceof JsonDocument &&
+        (document.url.endsWith('package.json') ||
+         document.url.endsWith('bower.json'))) {
+      this.packageFiles.push(document);
+    }
+  }
+
+  done() {
+    for (const packageFile of this.packageFiles) {
+      const dirname = path.dirname(packageFile.url);
+      if (!this.packagesByDir.has(dirname)) {
+        this.packagesByDir.set(
+            trimLeft(dirname, '/'), {schema_version: '1.0.0', elements: []});
+      }
+    }
+  }
+}
+
+class ElementGatherer implements AnalysisVisitor {
+  elements: Element[] = [];
+  private elementPaths = new Map<ElementDescriptor, string>();
+  visitElement(element: ElementDescriptor, path: Descriptor[]): void {
+    let pathToElement: string|null = null;
+    for (const descriptor of path) {
+      if (descriptor instanceof DocumentDescriptor) {
+        pathToElement = descriptor.document.url;
+      }
+    }
+    if (!pathToElement) {
+      throw new Error(`Unable to determine path to element: ${element}`);
+    }
+    if (this.elementPaths.has(element)) {
+      if (this.elementPaths.get(element) !== pathToElement) {
+        throw new Error(
+            `Found element ${element} at distinct paths: ` +
+            `${pathToElement} and ${this.elementPaths.get(element)}`);
+      }
+    } else {
+      this.elementPaths.set(element, pathToElement);
+      const elem = serializeElementDescriptor(element, pathToElement);
+      if (elem) {
+        this.elements.push(elem);
+      }
+    }
+  }
+}
+
+abstract class AnalysisVisitor {
+  visitDocumentDescriptor?(dd: DocumentDescriptor, path: Descriptor[]): void;
+  visitInlineDocumentDescriptor?
+      (dd: InlineDocumentDescriptor<any>, path: Descriptor[]): void;
+  visitElement?(element: ElementDescriptor, path: Descriptor[]): void;
+  visitDocument?(document: Document<any, any>, path: Descriptor[]): void;
+  done?(): void;
+}
+
+class AnalysisWalker {
+  private descriptors: DocumentDescriptor[];
+  private path: Descriptor[] = [];
+  constructor(descriptors: DocumentDescriptor[]) {
+    this.descriptors = descriptors;
+  }
+  walk(visitors: AnalysisVisitor[]) {
+    this.path.length = 0;
+    for (const descriptor of this.descriptors) {
+      this._walkDocumentDescriptor(descriptor, visitors);
+    }
+    for (const visitor of visitors) {
+      if (visitor.done) {
+        visitor.done();
+      }
+    }
+  }
+
+  private _walkDocumentDescriptor(
+      dd: DocumentDescriptor, visitors: AnalysisVisitor[]) {
+    this.path.push(dd);
+
+    for (const visitor of visitors) {
+      if (visitor.visitDocumentDescriptor) {
+        visitor.visitDocumentDescriptor(dd, this.path);
+      }
+    }
+
+    for (const entity of dd.entities) {
+      this._walkEntity(entity, visitors);
+    }
+    for (const dependency of dd.dependencies) {
+      this._walkEntity(dependency, visitors);
+    }
+    this._walkDocument(dd.document, visitors);
```

Comment:
As discussed, I _think_ we don't want to visit documents here, as they aren't really part of the descriptor tree. A visitor can still traverse into documents and their ASTs if needed though.